### ISSUE TITLE
isNative detection fails with NewRelic instrumentation

### ISSUE THREAD

**Issue by jdelStrother** [FezVrasta/popper.js#452]
### CodePen demo

https://codepen.io/jdelStrother/pen/LzrBXE

### Steps to reproduce the problem

1. Add NewRelic instrumentation, or any other instrumentation that wraps native functions (MutationObserver in particular)
2. Try to display something with Popper

### What is the expected behavior?

Popper uses MutationObserver and immediately displays the popped element at the correct position

### What went wrong?

Popper's isNative check fails, causing it to fall back to `setTimeout(fn, 0)`, causing the popped element to appear for a split second at the wrong position, before moving into place once the timeout fires.

### Any other comments?

`window.MutationObserver.toString = function() { return "[native code]" }` is a somewhat awful hack that works around Popper's isNative check.

**Comment by FezVrasta** [comment]
`isNative` already performs your exact same suggested check actually

**Comment by jdelStrother** [comment]
Mmm... mine isn't a check, it's overriding toString to satisfy Popper's isNative check.  Unless I misunderstand..?

**Comment by FezVrasta** [comment]
Ah sorry I didn't understand what you did there.

I don't think there's much to do if someone overrides the MutationObserver... we need to know if the MutationObserver is a native one or a polyfilled one.

**Comment by FezVrasta** [comment]
Maybe you want to tweak the fallback method (the one with `setTimeout`) to better work and avoid your problem?

PR welcome! But I'd like a reproducible example via CodePen to validate the problem.

**Comment by blainesch** [comment]
Blaine from NewRelic here. Is there a reason we can't use a polyfilled mutation observers? Presumably, they would use `setTimeout` so just detecting them would be fine?

**Comment by FezVrasta** [comment]
That's the reason why we can't use polyfills and why we fallback to a `setTimeout` based system in the case:
https://github.com/FezVrasta/popper.js/blob/08c5d6010346bf9df06e9f81a54fa6c2c51e3639/packages/popper/src/utils/debounce.js#L50-L53

**Comment by blainesch** [comment]
It looks like the only reason you use `isNative` is for mutation observer. Since we are worried about elements not connected to the dom, maybe we can do something like:

```
var isNative = false
var target = document.createElement('div')
var observer = new MutationObserver(function() {
  isNative = true
})
observer.observe(target, { attributes: true })
target.setAttribute('id', 'popper.js')
observer.disconnect()

module.exports = isNative
```

**Comment by FezVrasta** [comment]
The problem is that the polyfilled MutationObserver will not work properly on `target` because it's not attached to the DOM. (no `body.append(target)`)

**Comment by blainesch** [comment]
Right, so that should be a good test to see if you can use the mutation observer on the page or not without checking for `[native code]`, which should make this not have it flash positions? Sorry if I'm missing something.

**Comment by FezVrasta** [comment]
Maybe I miss something but, why a polyfilled `MutationObserver` shouldn't run the callback where you set `isNative` to `true`?

**Comment by blainesch** [comment]
New Relic (and other libraries) wraps `MutationObserver`, so from your point of view it isn't "native". This makes it fall back to the `setTimeout`. According to @jdelStrother, this makes this library lag for a few ms on the screen, even though `MutationObserver` is still native, it's just been wrapped by us.

The workaround, is to not check for `[native code]` but find another way to detect if `MutationObserver` is usable. The thing we need to find out is if `MutationObserver` can work if `target` isn't in the dom. The snippet I posted earlier seems to do this correctly, so it could be used instead of the current `isValid` method, correct?

**Comment by FezVrasta** [comment]
`target` will never be in the DOM. Because we generate this DOM element just to create a microTaskDebouncer.

**Comment by blainesch** [comment]
Exactly. In my snippet it's not in the DOM either. It should correctly detect if `MutationObserver` is usable by this library.

**Comment by jdelStrother** [comment]
Maybe the terminology is confusing the issue?   Blaine is suggesting something along these lines 

```diff
diff --git a/packages/popper/src/utils/debounce.js b/packages/popper/src/utils/debounce.js
index 53d986b..aaf029e 100644
--- a/packages/popper/src/utils/debounce.js
+++ b/packages/popper/src/utils/debounce.js
@@ -47,12 +47,21 @@ export function taskDebounce(fn) {
   };
 }
 
-// It's common for MutationObserver polyfills to be seen in the wild, however
-// these rely on Mutation Events which only occur when an element is connected
-// to the DOM. The algorithm used in this module does not use a connected element,
-// and so we must ensure that a *native* MutationObserver is available.
+
+function mutationObserverSupportsUnconnectedElements() {
+  var supportsUnconnected = false;
+  var target = document.createElement('div');
+  var observer = new MutationObserver(function() {
+    supportsUnconnected = true;
+  })
+  observer.observe(target, { attributes: true });
+  target.setAttribute('id', 'popper.js');
+  observer.disconnect();
+  return supportsUnconnected;
+}
+
 const supportsNativeMutationObserver =
-  isBrowser && isNative(window.MutationObserver);
+  isBrowser && mutationObserverSupportsUnconnectedElements()
 
 /**
 * Create a debounced version of a method, that's asynchronously deferred
```

ie we're not trying to determine if MutationObserver is native, just that it properly supports elements that aren't in the DOM.

Sadly, that snippet doesn't quite work, because the mutation observer callback is called asynchronously after `mutationObserverSupportsUnconnectedElements()` has exited .   We'd have to wait a tick to determine whether or not it's been called.  😕

**Comment by blainesch** [comment]
@jdelStrother thanks for the clarification. I'm not always great at explaining things [:

Yes, that is what I'm suggesting. If this approach is acceptable I'm sure we can find a way to make it work even with the async issue.

**Comment by FezVrasta** [comment]
I see, thanks for the clarification.

It looks like a good approach, if you manage to get a working PR I'll be happy to merge it.

**Comment by blainesch** [comment]
I can work on one. I'll have it up in a few minutes.

**Comment by blainesch** [comment]
@jdelStrother this was fixed with #461 and waiting on the next release.
